### PR TITLE
Space - Faezeh

### PR DESCRIPTION
**Pull request by faezeh-ashtiani**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | the spotlight method ins a class method on the Work model. it sorts all works based on their number of votes descending, if there is a tie, it picks the one that is alphabetically first, and then returns the first in the collection.
Describe how you approached testing that model method. What edge cases did you come up with? | I tested that is can pick the work with most votes, by having work features that are connected to vote features, and then by testing if it returns the one that is alphabetically first if there becomes a tie in voting. 
What are `session` and `flash`? What is the difference between them? | they are both hashes in rails. the flash has remains for the ongoing request cycle of the app and the next one. but the session hash remains in the app unless we eliminate it or close the browser.
What was one thing that you gained more clarity on through this assignment? | Oh so much! many to many relationships was one, testing was the other. view partials and so much more.
What is the Heroku URL of your deployed application? | https://fa-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+// body {
+//   margin: 0;
+//   // font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
+//   font-size: 1rem;
+//   font-weight: 400;
+//   line-height: 1.5;
+//   color: #212529;
+//   text-align: left;
+//   background-color: #fff;
+// }
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:38]: You forgot to remove some commented out code.

File: app/controllers/votes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        if @vote.save
+          flash[:success] = "Successfully upvoted!"
+        else
+          flash[:warning] = ["A problem occurred: Could not upvote", "user: has already voted for this work"]
+        end
+      end
+    else # no user logged in
+      flash[:warning] = "A problem occurred: You must log in to do that"
+    end
+    redirect_to request.referrer || works_path
```
Comment (kaidamasaki) [inline app/controllers/votes_controller.rb:21]: Style: It's cleaner to use `redirect_back` for this functionality:

```suggestion
    redirect_back(fallback_location: works_path)
```

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    @movies = Work.where(category: :movie).sort{ |a, b| b.votes.count <=> a.votes.count }
+    @books = Work.where(category: :book).sort{ |a, b| b.votes.count <=> a.votes.count }
+    @albums = Work.where(category: :album).sort{ |a, b| b.votes.count <=> a.votes.count }
+  end
+
+  def show
+    if @work.nil?
+      head :not_found
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:15]: Since you repeat this logic you can also move it into the controller filter.

Remember, with a controller filter if you return a response the action itself doesn't run so you don't need to worry about it continuing without a work.

File: app/views/homepages/_homepage.html.erb [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+          <h4>
+            <%= link_to "#{work.title}", work_path(work.id) %>
+            <small class="top-ten__creator">by <%= work.creator %></small>
+          </h4>
+          <p><%= work.votes.count %> votes</p>
+        </li>
+      <% end %>
+    </ul>
+    <!-- <a href="/works.movies">View More Movies</a> -->
+    </section>
```
Comment (kaidamasaki) [inline app/views/homepages/_homepage.html.erb:17]: Your indentation is a bit weird here.

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    <div class="<%= alert_class(flash) %>" > 
+      <% flash.each do |name, message| %>
+        <% unless message.respond_to?(:each) %>
+          <span><%= message %></span>
+        <% else %>
+          <span><%= message[0] %></span>
+          <ul>
+            <% message[(1..)].each do |error| %>
+              <li><%= message[1] %></li>
+            <% end %>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:61]: It's clearer to use `drop` here.  Also, I think you meant to use `error`:

```suggestion
            <% message.drop(1).each do |error| %>
              <li><%= error %></li>
            <% end %>
```

**Comment by kaidamasaki** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️?
| 17. Logging in as a new user creates a new user | ✔️?
| 18. Logging in as an already existing user has a specific flash message | ✔️?

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️?
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | No filter for finding a logged in user.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Your code was clean and clear and I especially liked how thorough your controller tests were.

There were a few little things you could fix but nothing major.  😃
